components/script/dom/bindings/construct_hook.rs

```diff
@@ -0,0 +1,59 @@
+// construct_hook.rs
+
+use std::ptr;
+use crate::dom::bindings::htmlconstructor;
+use crate::dom::bindings::utils::{GetConstructorNameForReporting, MakeNativeName};
+use crate::dom::bindings::root::DomRoot;
+use crate::jsapi::{JSContext, JSObject, CallArgs, GlobalScope, SafeJSContext};
+use crate::dom::bindings::import::module::callargs_is_constructing;
+use crate::dom::bindings::import::module::throw_constructor_without_new;
+use crate::dom::bindings::codegen::PrototypeList::ID;
+// Function for HTML constructor case
+pub unsafe fn construct_html_custom(global: &GlobalScope, cx: *mut JSContext, args: &CallArgs, descriptor_name: &str) -> bool {
+    let global = DomRoot::downcast::<dom::types::Global>(global).unwrap(); // Adjust type as needed
+    let prototype_id = match descriptor_name {
+        "HTMLDivElement" => PrototypeList::ID::HTMLDivElement,
+        "HTMLSpanElement" => PrototypeList::ID::HTMLSpanElement,
+        // Add cases for other descriptor names as needed
+        _ => panic!("Unsupported descriptor name: {}", descriptor_name),
+    };
+    htmlconstructor::call_html_constructor::<dom::types::Global>(
+        cx,
+        args,
+        &global,
+        prototype_id, // Adjust as needed
+        descriptor_name,
+    )
+}
+
+// Function for non-HTML constructor case
+pub unsafe fn construct_default_custom(global: &GlobalScope, cx: *mut JSContext, args: &CallArgs, descriptor_name: &str, ctor_name: &str) -> bool {
+    let prototype_id = match descriptor_name {
+        "HTMLDivElement" => PrototypeList::ID::HTMLDivElement,
+        "HTMLSpanElement" => PrototypeList::ID::HTMLSpanElement,
+        // Add cases for other descriptor names as needed
+        _ => panic!("Unsupported descriptor name: {}", descriptor_name),
+    };
+    if !callargs_is_constructing(args) {
+        throw_constructor_without_new(cx, ctor_name);
+        return false;
+    }
+
+    rooted!(in(cx) let mut desired_proto = ptr::null_mut::<JSObject>());
+    let proto_result = get_desired_proto(
+        cx,
+        args,
+        prototype_id, // Adjust as needed
+        descriptor_name,
+        desired_proto.handle_mut(),
+    );
+    assert!(proto_result.is_ok());
+    if proto_result.is_err() {
+        return false;
+    }
+
+    let nativeName = MakeNativeName(ctor_name);
```

Comment:
MakeNativeName is a python function. This is going to be a bit tricky; I think we're going to need to leave the actual constructor call in the generated python code, because there's a lot of unique logic in there. What we _can_ do is something like this:
```rust
pub unsafe fn construct_default_custom(..., constructor: impl FnOnce(*mut JSContext, &CallArgs, &GlobalScope) -> bool) -> bool {
  ...
  constructor(cx, args, global)
}
```
Then modify the python code to call `f"construct_default_custom(&global, cx, &args, \"{ctorName}\", |cx, args, global| {{ {constructorInvoke} }});\n"` where constructorInvoke is the result of calling `.define()` on the constructorCall CGMethodCall value.